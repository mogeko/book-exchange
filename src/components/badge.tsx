import { MessageContext } from "@/providers/msgProvider";
import { useContext } from "react";

const Badge: React.FC<BadgeProps> = ({ badgeKey, className }) => {
  const msg = useContext(MessageContext);
  const count = msg.filter((m) => m.key === badgeKey).length;
  return count ? <span className={`badge ${className}`}>{count}</span> : null;
};

interface BadgeProps {
  badgeKey: string;
  className?: string;
}

export default Badge;
