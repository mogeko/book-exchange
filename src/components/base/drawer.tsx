import { MenusContext } from "@/layouts/providers/menusProvider";
import Badge from "@/components/badge";
import { HiMenu } from "react-icons/hi";
import Link from "next/link";
import { useContext } from "react";

const DrawerMenu: React.FC<DrawerMenuProps> = ({ toggleId }) => {
  return (
    <div className="drawer-side">
      <label htmlFor={toggleId} className="drawer-overlay" />
      <MenusButtons />
    </div>
  );
};

const MenusButtons: React.FC = () => {
  const menus = useContext(MenusContext);

  return (
    <ul className="menu p-4 overflow-y-auto w-80 bg-base-200 text-base-content">
      {Object.entries(menus.root).map(([key, [name, href]], index) => (
        <li key={index}>
          <Link href={href}>
            <a className="justify-between">
              <span>{name}</span>
              <Badge badgeKey={key} />
            </a>
          </Link>
        </li>
      ))}
    </ul>
  );
};

export const withDrawer = (
  WrappedComponent: React.FC<Pick<DrawerProps, "children">>
) => {
  const DrawerWrapper: React.FC<DrawerProps> = (props) => {
    return (
      <div className="drawer">
        <input id={props.toggleId} type="checkbox" className="drawer-toggle" />
        <div className="drawer-content">
          <WrappedComponent>{props.children}</WrappedComponent>
        </div>
        <DrawerMenu toggleId={props.toggleId} />
      </div>
    );
  };
  return DrawerWrapper;
};

export const DrawerButton: React.FC<DrawerButtonProps> = ({ toggleId }) => {
  return (
    <label htmlFor={toggleId} className="btn btn-ghost btn-circle">
      <HiMenu className="h-6 w-6" aria-hidden="true" />
    </label>
  );
};

interface DrawerMenuProps {
  toggleId: string;
}

type DrawerProps = {
  children: React.ReactNode;
} & DrawerMenuProps;

type DrawerButtonProps = Pick<DrawerProps, "toggleId">;
