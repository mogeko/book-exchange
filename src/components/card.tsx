import { type BookTypes } from "@/lib/hooks/useBooks";
import Skeleton from "@/components/skeleton";
import StarsRate from "@/components/stars";
import Image from "next/image";
import Link from "next/link";

const CardRoot: React.FC<CardProps> = ({ title, cover, author, id }) => {
  return (
    <div className="relative flex flex-col w-28">
      <figure className="flex justify-center items-center">
        <Link href={`/books/${id}`}>
          <a className="relative w-28 h-[calc(7rem/2*3)]">
            <Image src={cover!} layout="fill" alt={title} />
          </a>
        </Link>
      </figure>
      <div className="flex flex-col text-sm">
        <Link href={`/books/${id}`}>
          <a className="link no-underline text-primary hover:bg-primary hover:text-primary-content focus:bg-primary-focus focus:text-primary-content">
            <h2 className="truncate">{title}</h2>
          </a>
        </Link>
        <p className="truncate">{author}</p>
      </div>
    </div>
  );
};

const CardSkeleton: React.FC = () => (
  <Skeleton.Pulse className="relative flex flex-col mb-4 w-28">
    <Skeleton.Square className="mb-2 w-28 h-[calc(7rem/2*3)]" />
    <Skeleton.Line className="mb-2 w-24" />
    <Skeleton.Line className="mb-2 w-20" />
  </Skeleton.Pulse>
);

const LongCardRoot: React.FC<CardProps> = (props) => {
  const { title, cover, author, id, description, rates } = props;
  return (
    <div className="flex py-4 gap-3">
      <figure className="flex justify-center items-center">
        <Link href={`/books/${id}`}>
          <a className="relative w-28 h-[calc(7rem/2*3)]">
            <Image src={cover!} layout="fill" alt={title} />
          </a>
        </Link>
      </figure>
      <div className="flex flex-col gap-2 h-[calc(7rem/2*3)]">
        <div className="inline-flex justify-start">
          <Link href={`/books/${id}`}>
            <a className="link no-underline text-primary hover:bg-primary hover:text-primary-content focus:bg-primary-focus focus:text-primary-content">
              <h2 className="truncate">{title}</h2>
            </a>
          </Link>
        </div>
        <p className="truncate">{author}</p>
        <StarsRate rates={rates!} />
        <p className="line-clamp-3 text-ellipsis">{description}</p>
      </div>
    </div>
  );
};

const LongCardSkeleton: React.FC = () => (
  <Skeleton.Pulse className="relative flex mb-4 w-full gap-3">
    <Skeleton.Square className="mb-2 w-28 h-[calc(7rem/2*3)]" />
    <div className="flex flex-col w-full gap-2">
      <Skeleton.Line className="mb-2 w-1/5" />
      <Skeleton.Line className="mb-2 w-1/6" />
      <div>
        <Skeleton.Line className="mb-2 w-4/5" />
        <Skeleton.Line className="mb-2 w-4/5" />
        <Skeleton.Line className="mb-2 w-4/5" />
        <Skeleton.Line className="mb-2 w-4/6" />
      </div>
    </div>
  </Skeleton.Pulse>
);

const Card = Object.assign(CardRoot, { Skeleton: CardSkeleton });
export const LongCard = Object.assign(LongCardRoot, {
  Skeleton: LongCardSkeleton,
});

type CardProps = BookTypes;

export default Card;
