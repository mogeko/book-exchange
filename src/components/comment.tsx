import Editor from "@/components/editor/editor";

const Comments: React.FC = () => {
  return <Editor />;
};

export default Comments;
