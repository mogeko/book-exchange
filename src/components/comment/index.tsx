import CommentCard from "@/components/comment/commentCard";
import useComments from "@/lib/hooks/useComments";
import Editor from "@/components/editor";
import fetcher from "@/lib/fetcher";
import type { EditorFormInput } from "@/components/editor";
import type { CommentType } from "@/lib/hooks/useComments";

const Comments: React.FC<CommentsProps> = ({ id }) => {
  const { data, isLoading } = useComments({ bkid: id });
  const handleSubmit = async (data: EditorFormInput) => {
    await fetcher("/api/comments/submit", {
      method: "POST",
      body: JSON.stringify(data),
      headers: {
        "Content-Type": "application/json",
      },
    });
  };

  return (
    <div className="flex flex-col gap-4 max-w-xl">
      <Editor onSubmit={handleSubmit} />
      {isLoading ? (
        <CommentCard.Skeleton />
      ) : (
        data?.map((comment: CommentType, index) => (
          <CommentCard key={index} comment={comment} />
        ))
      )}
    </div>
  );
};

interface CommentsProps {
  id: `bk${number}`;
}

export default Comments;
