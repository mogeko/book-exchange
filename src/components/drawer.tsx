import { HiMenu } from "react-icons/hi";
import Link from "next/link";
import useMenus from "@/lib/connect/menus";

const DrawerMenu: React.FC<DrawerMenuProps> = ({ toggleId }) => {
  const { menus, isLoading, isError } = useMenus("/api/menus/root.json");
  if (isLoading) return <div>Loading...</div>;
  if (isError) return <div>Error...</div>;
  return (
    <div className="drawer-side">
      <label htmlFor={toggleId} className="drawer-overlay" />
      <ul className="menu p-4 overflow-y-auto w-80 bg-base-200 text-base-content">
        {menus?.data.map((menu, index) => (
          <li key={index}>
            <Link href={menu.href}>
              <a className="justify-between">
                {menu.name}
                {menu.badge ? (
                  <span className="badge">{menu.badge}</span>
                ) : null}
              </a>
            </Link>
          </li>
        ))}
      </ul>
    </div>
  );
};

export const withDrawer = (
  WrappedComponent: React.FC<Pick<DrawerProps, "children">>
) => {
  const DrawerWrapper: React.FC<DrawerProps> = (props) => {
    return (
      <div className="drawer">
        <input id={props.toggleId} type="checkbox" className="drawer-toggle" />
        <div className="drawer-content">
          <WrappedComponent>{props.children}</WrappedComponent>
        </div>
        <DrawerMenu toggleId={props.toggleId} />
      </div>
    );
  };
  return DrawerWrapper;
};

export const DrawerButton: React.FC<DrawerButtonProps> = ({ toggleId }) => {
  return (
    <label htmlFor={toggleId} className="btn btn-ghost btn-circle">
      <HiMenu className="h-6 w-6" aria-hidden="true" />
    </label>
  );
};

interface DrawerMenuProps {
  toggleId: string;
}

type DrawerProps = {
  children: React.ReactNode;
} & DrawerMenuProps;

type DrawerButtonProps = Pick<DrawerProps, "toggleId">;
