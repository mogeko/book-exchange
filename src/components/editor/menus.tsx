import classNames from "@/lib/utils/classNames";
import { BubbleMenu as Bubble, FloatingMenu as Floating } from "@tiptap/react";
import { FaLink } from "react-icons/fa";
import { useRef, useState, useEffect } from "react";
import { useCallback, useMemo } from "react";
import type { ButtonProps } from "@/components/editor/buttons";
import type { Editor } from "@tiptap/react";

export const BubbleMenu: React.FC<MenuProps> = ({ editor, buttons }) => {
  const [isLinkActive, setIsLinkActive] = useState(false);
  const [hasLinkInput, setHasLinkInput] = useState(false);
  const inputRef = useRef<HTMLInputElement>(null);
  const handleSubmit = useCallback(() => {
    if (inputRef.current?.value) {
      editor?.chain().focus().setLink({ href: inputRef.current?.value }).run();
    } else {
      editor?.chain().focus().unsetLink().run();
    }
    setIsLinkActive(!isLinkActive);
  }, [editor, isLinkActive, setIsLinkActive]);
  const LinkDialog = useMemo(() => {
    return (
      <div className="input-group">
        <input
          onChange={(e) => setHasLinkInput(e.target.value !== "")}
          defaultValue={editor?.getAttributes("link").href}
          className="input input-bordered input-xs"
          ref={inputRef}
          autoFocus
        />
        <button
          onClick={handleSubmit}
          className={classNames(
            hasLinkInput ? "btn-primary" : "",
            "btn btn-xs"
          )}
        >
          OK
        </button>
      </div>
    );
  }, [editor, hasLinkInput, handleSubmit]);
  const LinkButton = useMemo(() => {
    return (
      <button
        onClick={() => setIsLinkActive(!isLinkActive)}
        className={classNames(
          editor?.isActive("link") ? "btn-active" : "",
          "btn btn-xs"
        )}
      >
        <FaLink className="w-3 h-3" />
      </button>
    );
  }, [editor, isLinkActive, setIsLinkActive]);

  useEffect(() => {
    editor?.isFocused && setIsLinkActive(false);
  }, [editor?.isFocused, setIsLinkActive]);

  return (
    editor && (
      <Bubble
        className="btn-group"
        tippyOptions={{ duration: 100 }}
        editor={editor}
      >
        {!isLinkActive &&
          buttons.map((Button, i) => <Button key={i} editor={editor} />)}
        {!isLinkActive && LinkButton}
        {isLinkActive && LinkDialog}
      </Bubble>
    )
  );
};

export const FloatingMenu: React.FC<MenuProps> = ({ editor, buttons }) => {
  return (
    editor && (
      <Floating
        className="btn-group"
        tippyOptions={{ duration: 100 }}
        editor={editor}
      >
        {buttons.map((Button, i) => (
          <Button key={i} editor={editor} />
        ))}
      </Floating>
    )
  );
};

interface MenuProps {
  editor: Editor | null;
  buttons: React.FC<ButtonProps>[];
}
