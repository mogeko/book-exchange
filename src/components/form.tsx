import { useForm } from "react-hook-form";
import type { UseFormRegister, UseFormProps } from "react-hook-form";
import type { FieldValues, SubmitHandler, FieldErrors } from "react-hook-form";
import React from "react";
import classNames from "@/lib/utils/classNames";

const Form: React.FC<FormProps> = (props) => {
  const { children, onSubmit, className, loading, opts } = props;
  const { handleSubmit, ...methods } = useForm(opts);
  const { errors } = methods.formState;

  return (
    <form className={className} onSubmit={handleSubmit(onSubmit)}>
      {React.Children.map(children, (child) => {
        if (!child.props.name && !child.props.label) return child;
        return React.createElement(child.type, {
          ...child.props,
          ...(loading ? { disabled: true } : {}),
          key: child.props.name ?? child.props.label,
          register: methods.register,
          errors: errors,
        });
      })}
    </form>
  );
};

export const Input: React.FC<InputProps> = (props) => {
  const { label, name, register, className, size = "md", ...rest } = props;
  const error = rest.errors?.[name ?? label?.toLowerCase()!]?.message;

  if (register && (name || label)) {
    return (
      <div className="form-control">
        {label && (
          <label className="label">
            <span className="label-text">{label}</span>
          </label>
        )}
        <input
          className={
            className ??
            classNames(
              "input input-bordered",
              error ? "input-error" : "",
              `input-${size}`
            )
          }
          placeholder={rest.placeholder ?? label?.toLowerCase() ?? name}
          {...register(name ?? label?.toLowerCase()!)}
          {...rest}
        />
        {error && (
          <label className="label">
            <span className="label-text-alt text-error">{error as any}</span>
          </label>
        )}
      </div>
    );
  }
  return <input className={className ?? "input input-bordered"} {...rest} />;
};

interface FormProps {
  children: React.ReactElement<FormChild>[];
  onSubmit: SubmitHandler<any>;
  opts?: UseFormProps<FieldValues>;
  className?: string;
  loading?: boolean;
}

type FormChild = {
  register?: UseFormRegister<FieldValues>;
  errors?: FieldErrors<FieldValues>;
  size?: "xs" | "sm" | "md" | "lg";
  label?: string;
  name?: string;
  className?: string;
};

type InputProps = FormChild & React.InputHTMLAttributes<HTMLInputElement>;

export default Form;
