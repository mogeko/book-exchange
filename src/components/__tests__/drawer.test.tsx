import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { DrawerButton, withDrawer } from "@/components/drawer";
import * as menus from "@/lib/connect/menus";

const ExampleDrawer = withDrawer((prop) => <div {...prop} />);

describe("User", () => {
  it("renders a globale menu", () => {
    render(<ExampleDrawer toggleId="menu-test">test</ExampleDrawer>);
    expect(screen.getByText("My library")).toBeInTheDocument();
  });

  it("snapshot a globale menu", () => {
    const { container } = render(
      <ExampleDrawer toggleId="menu-test">test</ExampleDrawer>
    );
    expect(container).toMatchSnapshot();
  });

  it("snapshot a DrawerButton", () => {
    const { container } = render(<DrawerButton toggleId="menu-test" />);
    expect(container.querySelector("svg")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
