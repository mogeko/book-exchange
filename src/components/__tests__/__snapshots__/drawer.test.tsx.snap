// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User snapshot a DrawerButton 1`] = `
<div>
  <label
    class="btn btn-ghost btn-circle"
    for="menu-test"
  >
    <svg
      aria-hidden="true"
      class="h-6 w-6"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 20 20"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M3 5a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM3 10a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM3 15a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1z"
        fill-rule="evenodd"
      />
    </svg>
  </label>
</div>
`;

exports[`User snapshot a globale menu 1`] = `
<div>
  <div
    class="drawer"
  >
    <input
      class="drawer-toggle"
      id="menu-test"
      type="checkbox"
    />
    <div
      class="drawer-content"
    >
      <div>
        test
      </div>
    </div>
    <div
      class="drawer-side"
    >
      <label
        class="drawer-overlay"
        for="menu-test"
      />
      <ul
        class="menu p-4 overflow-y-auto w-80 bg-base-200 text-base-content"
      >
        <li>
          <a
            href="/#"
          >
            My library
          </a>
        </li>
        <li>
          <a
            href="/#"
          >
            Ranking
          </a>
        </li>
        <li>
          <a
            href="/#"
          >
            Quote
          </a>
        </li>
        <li>
          <a
            href="/#"
          >
            I'm Feeling Lucky
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
